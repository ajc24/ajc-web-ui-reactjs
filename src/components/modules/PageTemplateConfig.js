/**
 * Developed by Anthony Cox in 2025
 */
import { storybookCssClassList } from '../data/storybook';

/* Commonly used CSS classes */
const bodyDefaultCss = 'body-default';

/**
 * Functionality designed to work with the page template components (headers, menu bars, body content, footers)
 */
export default class PageTemplateConfig {
  /**
   * Setup and configure the current web page to ensure the default CSS styling is set to it.
   * This also handles the styling of pages which are deployed on Storybook
   */
  static setupDocumentBodyCss() {
    /* If we are rendering on Storybook, remove the Storybook classes which will affect the width of the component */
    for (let index = 0; index < storybookCssClassList.length; index += 1) {
      if (document.body.classList.contains(storybookCssClassList[index]) === true) {
        document.body.classList.remove(storybookCssClassList[index]);
      }
    }
    /* Ensure that the <body> element stretches to 100% of the width of the screen, that the X axis scrollbar is hidden and Y axis scrollbar is always displayed */
    if (document.body.classList.contains(bodyDefaultCss) === false) {
      document.body.classList.add(bodyDefaultCss);
      document.activeElement.blur();
    }
  }
}
