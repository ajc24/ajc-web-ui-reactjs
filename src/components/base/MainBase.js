/**
 * Developed by Anthony Cox in 2024
 */
import React, { useEffect, useState } from 'react';
import PropTypes from 'prop-types';
import PageTemplateConfig from '../modules/PageTemplateConfig';
import '../css/common.css';
import './css/main-base.css';

const defaultComponentId = 'default-id--main-container';

/**
 * Baseline component setting the template for all Main components to be developed from.
 * This baseline auto-handles screen widths from the most commonly used mobile screen sizes (360x800) to the most commonly used desktop sizes (1920x1080).
 * The height of the component is 768px.
 */
const MainBase = props => {
  const [id, setId] = useState(defaultComponentId);

  useEffect(() => {
    /* Set the title for the current web page only if it has not already been set */
    document.title = props.title;
  
    /* Set the ID for the component */
    if (props.id !== undefined) {
      setId(`${props.id}--main-container`);
    }
    /* Setup the default CSS styling for the page which is rendering this component */
    PageTemplateConfig.setupDocumentBodyCss();
  });

  /* Set the CSS class lists for all DOM elements */
  const mainContainerCss = 'main-container background-white';
  const mainContentContainerCss = 'main-content-container background-transparent';

  return (
    <main id={id} className={mainContainerCss} aria-label={props.title}>
      <div className={mainContentContainerCss}>
        {props.children}
      </div>
    </main>
  );
}
MainBase.propTypes = {
  /** The content to be displayed within the main template component. */
  children: PropTypes.any,
  /** The unique identifier for this component. */
  id: PropTypes.string,
  /** The title of the web page content to be rendered in the main component. This title will be rendered in the browser tab. */
  title: PropTypes.string.isRequired,
};
export default MainBase;
