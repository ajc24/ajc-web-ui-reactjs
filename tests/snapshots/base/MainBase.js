/**
 * Developed by Anthony Cox in 2024
 */
import 'jsdom-global/register';
import React from 'react';
import { TestDev } from 'ajc-testing-code';
import MainBase from '../../../src/components/base/MainBase';
import PageTemplateConfig from '../../../src/components/modules/PageTemplateConfig';

describe('Main Base', () => {
  /* Set the spies for use in the tests */
  let setupDocumentBodyCssSpy;

  beforeAll(() => {
    setupDocumentBodyCssSpy = jest
      .spyOn(PageTemplateConfig, 'setupDocumentBodyCss')
      .mockImplementation(() => {});
  });

  afterAll(() => {
    setupDocumentBodyCssSpy.mockRestore();
  });

  describe('Component with default id', () => {
    let snapshot;
    
    beforeAll(() => {
      snapshot = TestDev.createSnapshot(
        <React.Fragment>
          <MainBase>
            Main base text content.
          </MainBase>
        </React.Fragment>
      );
    });

    it('verifies the snapshot for the component', () => {
      expect(snapshot).toMatchSnapshot();
    });
  });

  describe('Component with custom id', () => {
    let snapshot;
    
    beforeAll(() => {
      snapshot = TestDev.createSnapshot(
        <React.Fragment>
          <MainBase id="custom-id">
            Main base text content.
          </MainBase>
        </React.Fragment>
      );
    });

    it('verifies the snapshot for the component', () => {
      expect(snapshot).toMatchSnapshot();
    });
  });
});
