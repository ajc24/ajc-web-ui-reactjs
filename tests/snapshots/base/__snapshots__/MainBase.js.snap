// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Main Base Component with custom id verifies the snapshot for the component 1`] = `"<main id="custom-id--main-container" class="main-container background-white"><div class="main-content-container background-transparent">Main base text content.</div></main>"`;

exports[`Main Base Component with default id verifies the snapshot for the component 1`] = `"<main id="default-id--main-container" class="main-container background-white"><div class="main-content-container background-transparent">Main base text content.</div></main>"`;
